pkg/services/store/entity/sqlstash/folder_support.go

```diff
@@ -23,64 +26,109 @@ type folderInfo struct {
+		GUID:      fi.GUID,
+		Namespace: namespace,
+		UID:       fi.UID,
+		SlugPath:  fi.SlugPath,
+		JS:        js.String(),
+		Depth:     fi.depth,
+		Left:      fi.left,
+		Right:     fi.right,
+		Detached:  isLost,
+	})
+
+	for _, sub := range fi.children {
+		if err := sub.buildInsertItems(items, namespace, isLost); err != nil {
+			return nil
+		}
+	}
+
+	return nil
+}
+
+// This rebuilds the whole folders structure for a given namespace. This has to
+// be done each time an entity is created or deleted.
+// FIXME: This is very inefficient and time consuming. This could be implemented
+// with a different approach instead of MPTT, or at least mitigated by an async
+// job?
+// FIXME: This algorithm apparently allows lost trees which are called
+// "detached"? We should probably migrate to something safer.
+func (s *sqlEntityServer) updateFolderTree(ctx context.Context, x db.ContextExecer, namespace string) error {
+	_, err := x.ExecContext(ctx, "DELETE FROM entity_folder WHERE namespace=?", namespace)
 	if err != nil {
-		return err
+		return fmt.Errorf("clear entity_folder for namespace %q: %w", namespace, err)
 	}
 
-	query := "SELECT guid,name,folder,name,slug" +
-		" FROM entity" +
-		" WHERE " + s.dialect.Quote("group") + "=? AND resource=? AND namespace=?" +
-		" ORDER BY slug asc"
-	args := []interface{}{folder.GROUP, folder.RESOURCE, namespace}
+	listReq := sqlEntityListFolderElementsRequest{
+		SQLTemplate: sqltemplate.New(s.sqlDialect),
+		Group:       folder.GROUP,
+		Resource:    folder.RESOURCE,
+		Namespace:   namespace,
+		FolderInfo:  new(folderInfo),
+	}
+	query, err := sqltemplate.Execute(sqlEntityListFolderElements, listReq)
+	if err != nil {
+		return fmt.Errorf("execute SQL template to list folder items in namespace %q: %w", namespace, err)
+	}
 
-	all := []*folderInfo{}
-	rows, err := tx.Query(ctx, query, args...)
+	rows, err := x.QueryContext(ctx, query, listReq.GetArgs()...)
 	if err != nil {
-		return err
+		return fmt.Errorf("list folder items in namespace %q: %w", namespace, err)
 	}
-	defer func() { _ = rows.Close() }()
 
-	for rows.Next() {
-		folder := folderInfo{
-			children: []*folderInfo{},
-		}
-		err = rows.Scan(&folder.Guid, &folder.UID, &folder.parentUID, &folder.Name, &folder.Slug)
-		if err != nil {
-			return err
+	var itemList []*folderInfo
+	for i := 1; rows.Next(); i++ {
+		if err := rows.Scan(listReq.GetScanDest()...); err != nil {
+			return fmt.Errorf("scan row #%d listing folder items in namespace %q: %w", i, namespace, err)
 		}
-		all = append(all, &folder)
+		fi := *listReq.FolderInfo
+		itemList = append(itemList, &fi)
 	}
 
-	root, lost, err := buildFolderTree(all)
-	if err != nil {
-		return err
+	if err := rows.Close(); err != nil {
+		return fmt.Errorf("close rows after listing folder items in namespace %q: %w", namespace, err)
 	}
 
-	err = insertFolderInfo(ctx, tx, namespace, root, false)
+	root, lost, err := buildFolderTree(itemList)
```

Comment:
Mabye another FIXME.

Is this working with concurrent executions ? Two concurrent transactions might build a version of the folder tree that will not include created/deleted/updated items from the concurrent transaction. Is this assumption correct ?